metric/gauge.go

```diff
@@ -0,0 +1,208 @@
+type Gauge struct {
+	vals     sync.Map
+	desc     metricdata.Descriptor
+	start    time.Time
+	keys     []string
+	newEntry func() gaugeEntry
+}
+
+type gaugeEntry interface {
+	read(t time.Time) metricdata.Point
+}
+
+// Read returns the current values of the gauge as a metric for export.
+func (g *Gauge) Read() []*metricdata.Metric {
+	now := time.Now()
+	m := &metricdata.Metric{
+		Descriptor: g.desc,
+	}
+	g.vals.Range(func(k, v interface{}) bool {
+		entry := v.(gaugeEntry)
+		key := k.(string)
+		labelVals := g.labelValues(key)
+		m.TimeSeries = append(m.TimeSeries, &metricdata.TimeSeries{
+			StartTime:   g.start,
+			LabelValues: labelVals,
+			Points: []metricdata.Point{
+				entry.read(now),
+			},
+		})
+		return true
+	})
+	return []*metricdata.Metric{m}
+}
+
+func (g *Gauge) getEntry(labelVals []metricdata.LabelValue) gaugeEntry {
+	if len(labelVals) != len(g.keys) {
+		panic("must supply the same number of label values as keys used to construct this Gauge")
+	}
+	mapKey := g.mapKey(labelVals)
+	if entry, ok := g.vals.Load(mapKey); ok {
+		return entry.(gaugeEntry)
+	} else {
+		entry, _ := g.vals.LoadOrStore(mapKey, g.newEntry())
+		return entry.(gaugeEntry)
+	}
+}
+
+func (g *Gauge) mapKey(labelVals []metricdata.LabelValue) string {
+	vb := &tagencoding.Values{}
+	for _, v := range labelVals {
+		b := make([]byte, 1, len(v.Value)+1)
+		if v.Present {
+			b[0] = 1
+			b = append(b, []byte(v.Value)...)
+		}
+		vb.WriteValue(b)
+	}
+	return string(vb.Bytes())
+}
+
+func (g *Gauge) labelValues(s string) []metricdata.LabelValue {
+	vals := make([]metricdata.LabelValue, 0, len(g.keys))
+	vb := &tagencoding.Values{Buffer: []byte(s)}
+	for range g.keys {
+		v := vb.ReadValue()
+		if v[0] == 0 {
+			vals = append(vals, metricdata.LabelValue{})
+		} else {
+			vals = append(vals, metricdata.NewLabelValue(string(v[1:])))
+		}
+	}
+	return vals
+}
+
+// Float64Gauge represents a float64 value that can go up and down.
+//
+// Float64Gauge maintains a float64 value for each combination of of label values
+// passed to the Set or Add methods.
+type Float64Gauge struct {
+	Gauge
+}
+
+type float64Entry struct {
+	val uint64
+}
+
+func (e *float64Entry) read(t time.Time) metricdata.Point {
+	return metricdata.NewFloat64Point(t, math.Float64frombits(atomic.LoadUint64(&e.val)))
+}
+
+// NewFloat64WithRegistry creates a new gauge with a float64 value.
+func NewFloat64Gauge(name, description string, unit metricdata.Unit, keys ...string) *Float64Gauge {
+	g := &Float64Gauge{
+		Gauge{
+			newEntry: func() gaugeEntry {
+				return new(float64Entry)
+			},
+			keys:  keys,
+			start: time.Now(),
+			desc: metricdata.Descriptor{
+				Name:        name,
+				Description: description,
+				Unit:        unit,
+				LabelKeys:   keys,
+			},
+		},
+	}
+	return g
+}
+
+// Set sets the current gauge value.
+func (g *Float64Gauge) Set(val float64, labelVals ...metricdata.LabelValue) {
+	ge := g.getEntry(labelVals).(*float64Entry)
```

Comment:
getEntry seems to be expensive:
* need to serialize the entries
* possible lock contention on the sync.Map

This were the reason we exposed the "Entry" (in java LongPoint) to the user to allow them to cache the entry and avoid all the possible performance implications.

Reply:
I thought it would be worth not exposing it because it would make the API more complex. I will change it to expose it directly.